### ISSUE TITLE
[rl] PPG: `aux_buffer`/`ppg_agent` per-obs device round-trip; use `write_host_row`, not `Tensor::stack`

### ISSUE THREAD

**Issue by anthonytorlucci** [anthonytorlucci/rlevo#187]
> **Rewritten and rescoped 2026-07-17** after re-triage against the current tree. The perf issue is real but misdiagnosed, **the recommended fix would make it worse**, the encapsulation claim is refuted, and the larger instance of the same defect was missed. `priority: high` removed. Original text is in the edit history. See "Triage record" below.

## The perf bug (was §2.1)

Not "re-uploads one at a time" — the final upload **is** batched once (`aux_buffer.rs:152-153`). The real waste is a per-observation **upload-then-download**:

```rust
slice.obs[ii].to_tensor(device)   // aux_buffer.rs:143 — pushes one obs to the device
             .into_data()          // aux_buffer.rs:144 — immediately pulls it back
```

That is `2n` device transfers per minibatch, purely to obtain a host-side `Vec<f32>` conversion.

### `Tensor::stack` is the wrong fix

`TensorConvertible` already exposes a host-side seam: `write_host_row(&self, buf: &mut Vec<f32>)` (`rlevo-core/src/base.rs:224`). `to_tensor` is a *provided* method built on it and is marked **"# Do not override"** (`base.rs:226-240`).

Correct fix: `slice.obs[ii].write_host_row(&mut obs_flat)` — zero device touches before the single batched upload. **`Tensor::stack` would retain every per-obs upload and add a stack op on top** — strictly worse than the status quo.

### The bigger instance, missed by the original

`ppg_agent.rs:750-754` repeats the **exact same** `to_tensor` → `into_data()` round-trip in the π_old logits path — over **`total_steps`** (the whole pool), not just a minibatch. That is strictly the larger cost of the two. **Any fix here must land there too**, or this issue fixes the cheaper half and leaves the expensive one.

## The doc bug (was §3.1, rescoped)

The cap claim is **not** on `AuxRolloutSlice` (its doc at `aux_buffer.rs:30` claims nothing about caps). It is the **module doc at `aux_buffer.rs:3`**:

> *"Accumulates the last `n_iteration` policy-phase rollouts"*

The word **"last"** implies a bounded window. `push_slice` (`:66`) has no cap and never evicts. In the normal loop this is masked because `maybe_aux_phase` drains via `clear()` (`ppg_agent.rs:719`, early-return at `:649`) — but `snapshot_into_aux_buffer` is `pub` (`ppg_agent.rs:478`), so a caller driving it without the aux phase grows the pool unboundedly.

**Prefer fixing the doc** ("accumulates until drained by the auxiliary phase") over adding a cap. The code's behaviour is correct for its actual use; the doc describes a different design.

---

## Triage record

| Original claim | Verdict |
|---|---|
| §2.1 `gather_minibatch` re-uploads observations one at a time | **PARTIAL** — round-trip real, mechanism wrong; the final upload *is* batched (`:152-153`) |
| §2.1 fix: use `Tensor::stack` | **REFUTED — wrong fix**, would retain the round-trips and add work. Use `write_host_row`. |
| §3.1 `AuxRolloutSlice` has `pub` fields | **CONFIRMED** (`aux_buffer.rs:33-36`) |
| §3.1 the `pub` fields **break its own invariant** | **REFUTED** — `AuxRolloutBuffer::slices` is **private** (`:42`) and `AuxRolloutSlice` is only ever constructed inside `push_slice`, which asserts `returns.len() == obs.len()` (`:67-73`). No setter admits an externally-built slice, so the `pub` fields cannot violate the length invariant. Cosmetic. |
| §3.1 doc-claims a cap invariant it doesn't enforce | **RESCOPED** — real, but at `:3` (module doc), not on the slice; fix the doc |
| Severity: High | **RESCOPED → Medium** |

## Related

- #172 / #322 — the same `to_tensor` → `into_data` round-trip pattern elsewhere in the crate; consider one coordinated fix
